### DISCUSSION TITLE
Add support for a test-timeout option

### DISCUSSION THREAD

**Discussion by CleanCut** [CleanCut/green#45] [Enhancements]
On StackOverflow, [it was asked](http://stackoverflow.com/questions/29311096/timing-out-tests-with-green-unit-test-runner) whether green had any timeout functionality similar to the timeout plugin in py.test.  Sounds like an interesting feature.

What would be the best behavior when a test hits the timeout?
- Count as "pass"
- Count as "fail"
- Count as "skipped"
- Count as "expected failure"
- Something else?

**Comment by CleanCut** [comment]
Ooh, or maybe the timeout behavior should be configurable too.  Something like

```
green --timeout=200 --timeout-result=fail
```

**Comment by miohtama** [comment]
I think for the inspiration we can look here:

https://bitbucket.org/flub/pytest-timeout/
- Timeout should be counted as a failure (think exception raised inside the test). By the default, tests should never timeout. If you want to skip long running tests you can do it using the skip decorator.
- You can specify timeout from command line, to make sure the test suite finishes timely
- You can also specify one using a function decorator, esp. if you know the test is doing third party service API calls which may hang

From py.test:
- Timeoutting is done using threading.Timer and sending signal.alarm 

There seems to be also some really complex corner cases, like entering pdb: https://bitbucket.org/flub/pytest-timeout/src/f97e9c8b79a4da9b89b49fb67be91f6c3c27457f/pytest_timeout.py?at=default#cl-82

Because py.test does not support timeout natively, but everything is done as a plugin hack, I guess one can come up with a cleaner implementation. 

Also it probably does not make sense to have perfect cleanup or teardown in the case of timeout. Think it more of an valve to make sure the test run terminates in some point and does not keep running forever unless there is human intervention.

**Comment by CleanCut** [comment]
I'm not interested in implementing this myself.  If someone else wants to, I would be happy to review the pull request.

**Comment by CleanCut** [comment]
I'm going to go ahead and re-open this.  Even if I don't implement it personally (or at least don't do it _soon_), it's still a good enhancement idea in general.
